### DISCUSSION TITLE
Training batch size and adaptive random forest graph, feature importance

### DISCUSSION THREAD

**Discussion by IvanLiuTW** [online-ml/river#1051] [Q&A] [answered]
Hi all,

I am working on a regression project, and the data I generated is time series data. The data was collected on a daily basis, so I want to train my model when I have a specific amount of data, or the data might be too less to train a good model. Since I'd like to set a minimum amout of data to retrain the model, I am wondering if there is any parameter like **scikit-multiflow** can be set in a function, like this:

```
evaluator = EvaluatePrequential(
    pretrain_size=30, # Set the size of the pretrain data
    batch_size = 30, # Set the size of data used to retrain the model
    max_samples=623,
    metrics = ['mean_absolute_error', 'true_vs_predicted'],
    output_file='results.csv')
```

On the other hand, I found in documention that the we can draw the tree structure from `Hoeffding Tree` model. Can the same method applied to the `Adaptive Random Forest`? And can I get feature importance from `Adaptive Random Forest Regressor`, just like we get from `sklearn's Random Forest Regressor`?

Thank you!

**Comment by MaxHalford** [comment, marked as answer]
Hey there. Ok so you have three questions :)

> I am working on a regression project, and the data I generated is time series data. The data was collected on a daily basis, so I want to train my model when I have a specific amount of data, or the data might be too less to train a good model. Since I'd like to set a minimum amout of data to retrain the model, I am wondering if there is any parameter like scikit-multiflow can be set in a function, like this:

IIRC, you want to update your model once you've collected a batch of sufficient size, correct? If so, there's no canonical way to do this with River. This is mostly because we don't work support mini-batches as much as scikit-multiflow does. But I see no reason why you shouldn't be able to implement this yourself. It sounds a tad too specific to your use case, so I doubt we should add support for this in River.

> On the other hand, I found in documention that the we can draw the tree structure from Hoeffding Tree model. Can the same method applied to the Adaptive Random Forest?

You should be able to draw each tree individually. Something like this should work:

```py
forest = AdaptiveRandomForestClassifier()
# ...
forest[0].draw()
```

> And can I get feature importance from Adaptive Random Forest Regressor, just like we get from sklearn's Random Forest Regressor?

Alas we don't have this right now. I'll let @smastelini confirm. I think that in general we need to think of a general framework for expressing feature importances. We're a bit fragile on this aspect.

**Reply by smastelini** [reply]
Sorry for my late reply. 

Indeed, we do not have support to feature importance. I've seen a few works start addressing this kind of task, but this is still in the research realm. I am not aware of a canonical way of calculating feature importance out of incremental decision tree forests.

**Reply by MaxHalford** [reply]
IIRC there are some methods that simply aggregate the branch and leaf impurities.

**Reply by IvanLiuTW** [reply]
Hi @MaxHalford  and @smastelini,

Thanks for reply and the suggestion! : )

> But I see no reason why you shouldn't be able to implement this yourself. It sounds a tad too specific to your use case, so I doubt we should add support for this in River.

I understand, and I did actually think about implementing this myself. Since I know River from Scikit-multiflow's Github, and they mentioned they are working on integrating scikit-multiflow with River, I was wondering if River would have the same feature as scikit-multiflow did.

If I would like to keep using river, could you please take a look of my method and I would be appreciated if you could give me any suggestion. I think of one way to do this:

1. I split my data into specific size of subdata, and train the model
```
for i in range(0, 30):
    # Load data
    data = iter_csv('stream_data_'+str(i)+.'csv', target = 'nextDayRestHR', converters=param)
    # Create model
    model = (
        StandardScaler() |
        ensemble.AdaptiveRandomForestRegressor(n_models = 100, seed = 18)
    )
    
    metric = metrics.MAE()
    evaluate.progressive_val_score(data, model, metric)
```

> Indeed, we do not have support to feature importance. I've seen a few works start addressing this kind of task, but this is still in the research realm. I am not aware of a canonical way of calculating feature importance out of incremental decision tree forests.

I see. Look forward to this feature!

Thanks Max and Saulo for answering my question.

**Reply by MaxHalford** [reply]
Hey there @IvanLiuTW.

So I understand that you're reading your data in mini-batches, correct? In the your code, you reinitialize your model for each mini-batch, right? Is there any particular reason you're doing that?

I'm not sure I understand why you're not doing something simpler, which is simply to loop through all your data, sample by sample, and learn with each sample. What exactly are you worried about?

**Reply by IvanLiuTW** [reply]
Hi Max,

Thanks for reply and pointing out the problem. I will try what you suggested.
